### PR TITLE
Updated to Support Reading/Writing to/from Streams

### PR DESCRIPTION
**Pull request by dellis1972**
Added bits for Windows Support (needs reviewing)
Added more Tests to the ZipTest Program
Added a copy of the ZipExtensions that we use in Xamarin.Android
to the ZipTest. We might want to look into adding these to the API?

### CODE REVIEW

File: ZipArchive.cs [hunk 1]
```diff
@@ -255,13 +440,16 @@ string NormalizeArchivePath (bool isDir, string archivePath)
...
+					archivePath = archivePath.Substring (0, archivePath.Length - 1);
+
+			if (Path.IsPathRooted (archivePath)) {
+				archivePath = archivePath.Remove (0, Path.GetPathRoot (archivePath).Length);
 			}
 
-			if (archivePath [archivePath.Length - 1] == Path.DirectorySeparatorChar)
-				return archivePath.Substring (0, archivePath.Length - 1);
-			return archivePath;
+			return archivePath.Replace ("\\", "/");
```
Comment (dellis1972) [inline ZipArchive.cs]: We kinda need this because users might give us pass data with "\" rather than "/"

File: ZipTest/Program.cs [hunk 2]
```diff
@@ -63,12 +65,123 @@ public static void Main (string [] args)
...
+	}
+
+	public enum CompressionLevel
+	{
+		NoCompression, 
+		Optimal,
+		Fastest,
+	}
+
+	static class ZipArchiveExtensions
```
Comment (grendello) [inline ZipTest/Program.cs]: You're right that those extensions should be added to the API. Can you put them directly in ZipArchive?

**Comment by grendello** [comment]
Please move ZipArchiveExtensions directly to ZipArchive, add /// docs and then it LGTM :)

File: ZipArchive.cs (hunk 1 above)
Comment (grendello) [inline ZipArchive.cs]: Agreed

File: ZipArchive.cs [hunk 3]
```diff
@@ -248,20 +376,115 @@ public ZipEntry AddFile (string sourcePath, string archivePath = null, EntryPerm
...
+			AddStream (data, entryName, method: method);
+		}
+
+		/// <summary>
+		/// Adds the byte[] data to the archive
+		/// </summary>
+		/// <param name="entryName">The name of the entry with in the archive</param>
+		/// <param name="data">The draw byte[] array to add to the entry</param>
+		/// <param name="method">The compression method to use</param>
+		public void AddEntry (string entryName, byte [] data,
```
Comment (grendello) [inline ZipArchive.cs]: I don't think we need this overload. The existing `AddEntry` for `byte[]` is good enough.

File: ZipArchive.cs [hunk 4]
```diff
@@ -248,20 +376,115 @@ public ZipEntry AddFile (string sourcePath, string archivePath = null, EntryPerm
...
+		/// the raw file.
+		/// </summary>
+		/// <param name="entryName">The name of the entry with in the archive</param>
+		/// <param name="text">The text to add to the entry</param>
+		/// <param name="encoding">The Encoding to use for the data.</param>
+		/// <param name="method">The compression method to use</param>
+		public void AddEntry (string entryName, string text, Encoding encoding,
+			CompressionMethod method = CompressionMethod.DEFAULT)
+		{
+			AddEntry (entryName, encoding.GetBytes (text), method);
```
Comment (grendello) [inline ZipArchive.cs]: Add null checks. `encoding` must not be null (or, better, should default to `Encoding.Default`) and passing null string to `GetBytes` will throw an error

File: ZipArchive.cs [hunk 5]
```diff
@@ -248,20 +376,115 @@ public ZipEntry AddFile (string sourcePath, string archivePath = null, EntryPerm
...
+		}
+
+		/// <summary>
+		/// Add a list of files to the archive
+		/// </summary>
+		/// <param name="fileNames">An IEnumerable<stirng> of files to add</param>
+		/// <param name="directoryPathInZip">The root directory path in archive.</param>
+		public void AddFiles (IEnumerable<string> fileNames, string directoryPathInZip)
+		{
+			foreach (var file in fileNames) {
```
Comment (grendello) [inline ZipArchive.cs]: Add a null check for `fileNames`

File: ZipArchive.cs [hunk 6]
```diff
@@ -248,20 +376,115 @@ public ZipEntry AddFile (string sourcePath, string archivePath = null, EntryPerm
...
+
+		/// <summary>
+		/// Recursively Add an entire directory structure to the archive
+		/// </summary>
+		/// <param name="folder">The root of the directory to add</param>
+		/// <param name="folderInArchive">The root name of the folder in the zip.</param>
+		/// <param name="method">The compresison method to use when adding files</param>
+		public void AddDirectory (string folder, string folderInArchive, CompressionMethod method = CompressionMethod.DEFAULT)
+		{
+			string root = folderInArchive;
```
Comment (grendello) [inline ZipArchive.cs]: Add null checks. Passing a `null` to `Directory.GetFiles` throws an exception. Also, `folderInArchive` should be required to be non-null

File: ZipArchive.cs [hunk 7]
```diff
@@ -248,20 +376,115 @@ public ZipEntry AddFile (string sourcePath, string archivePath = null, EntryPerm
...
+			}
+			foreach (var dir in Directory.GetDirectories (folder)) {
+				var internalDir = dir.Replace ("./", string.Empty).Replace (folder, string.Empty);
+				AddDirectory (dir, folderInArchive + internalDir, method);
+			}
+		}
+
+		string ArchiveNameForFile (string filename, string directoryPathInZip)
+		{
+			string pathName;
```
Comment (grendello) [inline ZipArchive.cs]: Add a null check for `filename`

File: ZipArchive.cs (hunk 4 above)
Comment (dellis1972) [inline ZipArchive.cs]: I'll add null checks. We cannot use Encoding.Default because its not a compile time constant :(

Comment (grendello) [inline ZipArchive.cs]: You do it in the body of the method:

`encoding = encoding ?? Encoding.Default`

or 

`(encoding ?? Encoding.Default).GetBytes (text)`

**Pull request merged**